### ISSUE TITLE
Provisioning "changed" to the current value

### ISSUE THREAD

**Issue by sebdah** [sebdah/dynamic-dynamodb#277]
It's not allowed to change provisioning from and to the same value. So changing read throughput from `1` to `1` will throw an error in the AWS API. This bug does precisely that. See the report from @ankushsachdeva in #28.

He writes:

> I am having this issue on v2.1.0 with GSI
> ```
ubuntu@ip-xxx:~/dynamic-dynamodb$ ./dynamic-dynamodb -c dynamic-dynamodb.conf 
2015-10-04 20:35:33,226 - dynamic-dynamodb - INFO - tagPosts - Consumed read units: 0.00%
2015-10-04 20:35:33,706 - dynamic-dynamodb - INFO - tagPosts - Consumed write units: 0.00%
2015-10-04 20:35:34,463 - dynamic-dynamodb - INFO - tagPosts - Consumed read units: 0.00%
2015-10-04 20:35:34,697 - dynamic-dynamodb - INFO - tagPosts - Read throttle count: 0
2015-10-04 20:35:35,173 - dynamic-dynamodb - INFO - tagPosts - Throttled read percent by provision: 0.00%
2015-10-04 20:35:35,638 - dynamic-dynamodb - INFO - tagPosts - Throttled read percent by consumption: 0.00%
2015-10-04 20:35:35,638 - dynamic-dynamodb - INFO - tagPosts - Scaling down reads is not done when usage is at 0%
2015-10-04 20:35:35,638 - dynamic-dynamodb - INFO - tagPosts - Consecutive read checks 0/1
2015-10-04 20:35:36,343 - dynamic-dynamodb - INFO - tagPosts - Consumed write units: 0.00%
2015-10-04 20:35:36,601 - dynamic-dynamodb - INFO - tagPosts - Write throttle count: 0
2015-10-04 20:35:37,070 - dynamic-dynamodb - INFO - tagPosts - Throttled write percent by provision: 0.00%
2015-10-04 20:35:37,547 - dynamic-dynamodb - INFO - tagPosts - Throttled write percent by consumption: 0.00%
2015-10-04 20:35:37,547 - dynamic-dynamodb - INFO - tagPosts - Scaling down writes is not done when usage is at 0%
2015-10-04 20:35:37,547 - dynamic-dynamodb - INFO - tagPosts - Consecutive write checks 0/1
2015-10-04 20:35:37,548 - dynamic-dynamodb - INFO - tagPosts - No need to change provisioning
2015-10-04 20:35:37,795 - dynamic-dynamodb - INFO - tagPosts - Will ensure provisioning for global secondary index tagId-postPostedOn-index
2015-10-04 20:35:38,272 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Consumed read units: 0.00%
2015-10-04 20:35:38,747 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Consumed write units: 0.00%
2015-10-04 20:35:39,458 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Consumed read units: 0.00%
2015-10-04 20:35:39,701 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Read throttle count: 0
2015-10-04 20:35:40,176 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Throttled read percent by provision: 0.00%
2015-10-04 20:35:40,653 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Throttled read percent by consumption: 0.00%
2015-10-04 20:35:40,654 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Increasing reads tomeet gsi-min-provisioned-reads limit (10 reads)
2015-10-04 20:35:40,654 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Consumed is over proposed read units. Will leave table at current setting.
2015-10-04 20:35:40,654 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Consecutive read checks 0/1
2015-10-04 20:35:41,360 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Consumed write units: 0.00%
2015-10-04 20:35:41,610 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Write throttle count: 0
2015-10-04 20:35:42,090 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Throttled write percent by provision: 0.00%
2015-10-04 20:35:42,587 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Throttled write percent by consumption: 0.00%
2015-10-04 20:35:42,587 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Increasing writes tomeet gsi-min-provisioned-writes limit (10 writes)
2015-10-04 20:35:42,587 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Consumed is over proposed write units. Will leave table at current setting.
2015-10-04 20:35:42,587 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Consecutive write checks 0/1
2015-10-04 20:35:42,587 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Changing provisioning to 1.000000 read units and 1.000000 write units
2015-10-04 20:35:43,778 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Updating provisioning to 1 reads and 1 writes
2015-10-04 20:35:44,015 - dynamic-dynamodb - ERROR - tagPosts - GSI: tagId-postPostedOn-index - Unhandled exception: ValidationException: The provisioned throughput for the index tagId-postPostedOn-index will not change. The requested value equals the current value. Current ReadCapacityUnits provisioned for index tagId-postPostedOn-index: 1. Requested ReadCapacityUnits: 1. Current WriteCapacityUnits provisioned for index tagId-postPostedOn-index: 1. Requested WriteCapacityUnits: 1. Refer to the Amazon DynamoDB Developer Guide for current limits and how to request higher limits.. Please file a bug report at https://github.com/sebdah/dynamic-dynamodb/issues
```

I'm seeing one odd thing in this, namely the `2015-10-04 20:35:42,587 - dynamic-dynamodb - INFO - tagPosts - GSI: tagId-postPostedOn-index - Consumed is over proposed write units. Will leave table at current setting.`. Firstly I don't really understand what the purpose of this check is. But I guess @Sazpaimon can shed some light on that.

Secondly it seems like `update_needed` is set to `False` when the `__ensure_*_provisioning` function is complete. But still the provisioning is changed.

**Comment by sebdah** [comment]
@Sazpaimon I'm thinking that maybe we should move https://github.com/sebdah/dynamic-dynamodb/blob/master/dynamic_dynamodb/core/gsi.py#L566 above the ensure min/max if statements (https://github.com/sebdah/dynamic-dynamodb/blob/master/dynamic_dynamodb/core/gsi.py#L539). It seems strange to not have the min / max checks last.

What is your ideas around this?

**Comment by Sazpaimon** [comment]
The problem with putting it before the max would be if the max is actually under what is currently being consumed. It would cause the autoscaler to lower the throughput to be under the current consumed level. This could be due to, say an external process that is increasing the throughput of a table so it can be extracted via an ETL process. I intended it so when we have a situation like this, have the autoscaler do nothing, hence setting `update_needed` to `False`. That's really strange that the throughput is still being updated, though. I only tested this on regular tables though, since I'm currently not using GSIs, so I don't know if I screwed something up there (it's possible :( )

**Comment by Sazpaimon** [comment]
Looking at it further, it looks like this may have been fixed by #271, as the problem was `update_needed = False` was set at the right indentation level, but the two lines under it were indented 1 level in, causing them to take place outside of the if statement

**Comment by sebdah** [comment]
You're up early :). Or late?

Thanks for the clarification, @Sazpaimon! So @ankushsachdeva, try bumping to the latest release of Dynamic DynamoDB and verify that the issue is gone. Please report back here with the results.
